task/snitch/src/main.rs

```diff
@@ -0,0 +1,167 @@
+use task_net_api::{
+    LargePayloadBehavior, Net, RecvError, SendError, SocketName,
+};
+use task_packrat_api::Packrat;
+use userlib::{sys_recv_notification, task_slot};
+use zerocopy::TryFromBytes;
+
+task_slot!(NET, net);
+task_slot!(PACKRAT, packrat);
+
+#[derive(Count, Copy, Clone)]
+enum Event {
+    RecvPacket,
+    RequestRejected,
+    ReadError(#[count(children)] task_packrat_api::EreportReadError),
+    Respond,
+}
+
+struct StaticBufs {
+    rx_buf: [u8; REQ_SZ],
+    tx_buf: [u8; UDP_PACKET_SZ],
+}
+
+const REQ_SZ: usize = core::mem::size_of::<Request>();
+const UDP_PACKET_SZ: usize = 1024;
+
+counters!(Event);
+
+#[export_name = "main"]
+fn main() -> ! {
+    let net = Net::from(NET.get_task_id());
+    let packrat = Packrat::from(PACKRAT.get_task_id());
+
+    const SOCKET: SocketName = SocketName::ereport;
+
+    let StaticBufs {
+        ref mut rx_buf,
+        ref mut tx_buf,
+    } = {
+        static BUFS: static_cell::ClaimOnceCell<StaticBufs> =
+            static_cell::ClaimOnceCell::new(StaticBufs {
+                rx_buf: [0u8; REQ_SZ],
+                tx_buf: [0u8; UDP_PACKET_SZ],
+            });
+        BUFS.claim()
+    };
+
+    loop {
+        let meta = match net.recv_packet(
+            SOCKET,
+            LargePayloadBehavior::Discard,
+            &mut rx_buf[..],
+        ) {
+            Ok(meta) => meta,
+            Err(RecvError::QueueEmpty) => {
+                // Our incoming queue is empty. Wait for more packets.
+                sys_recv_notification(notifications::SOCKET_MASK);
+                continue;
+            }
+            Err(RecvError::ServerRestarted) => {
+                // `net` restarted; just retry.
+                continue;
+            }
+        };
+
+        // Okay, we got a packet!
+        count!(Event::RecvPacket);
+        let request =
+            match Request::try_ref_from_bytes(&rx_buf[..meta.size as usize]) {
+                Ok(req) => req,
+                Err(_) => {
+                    // We ignore malformatted, truncated, etc. packets.
+                    count!(Event::RequestRejected);
+                    continue;
+                }
+            };
+
+        let size = match request {
+            Request::V0(req) => match packrat.read_ereports(
+                req.request_id,
+                req.restart_id,
+                req.start_ena,
+                req.limit,
+                req.committed_ena()
+                    .copied()
+                    .unwrap_or(gateway_ereport_messages::Ena::NONE),
+                &mut tx_buf[..],
+            ) {
+                Ok(size) => size,
+                Err(e) => {
+                    // Packrat's mad. Reject the request.
+                    //
+                    // Presently, the only time we'd see an error here is if we
+                    // have yet to generate a restart ID.
+                    count!(Event::ReadError(e));
+                    continue;
```

Comment:
It occurs to me that another possible way to handle this state is to have the snitch wait for a notification from packrat saying that the restart ID has been set, *before* it tries to handle any incoming packets. I'm not sure if there's any real advantage to letting the packets queue up rather than dropping them, though, because if we let them queue up, MGS will time them out and retry them anyway. So I think this ends up looking pretty much equivalent.

Reply:
It would take some priority shuffling, but one way to recover from `packrat` restarts is to have `snitch` poke the RNG driver to deliver a new restart ID to packrat here!

This would require `snitch -> rng driver -> packrat` in terms of priorities, so someone could theoretically starve snitch by... spamming the RNG driver with requests (seems unlikely).